apps/examples/araplayer/AudioPDStream.cpp

```diff
@@ -0,0 +1,192 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+#include "AudioPDStream.h"
+#include "MediaPlayerWrapper.h"
+
+AudioPDStream::AudioPDStream()
+{
+}
+
+AudioPDStream::~AudioPDStream()
+{
+	removeMapAll();
+}
+
+int AudioPDStream::startStream()
+{
+	AudioHttpDownloader *pHttpDownloader;
+	int ret = -1;
+
+	pHttpDownloader = new AudioHttpDownloader();
+	if (pHttpDownloader == NULL) {
+		DEBUG_TRACE(LOG_ERR, "MultiRoomHttpDownloader error!");
+		return -1;
+	}
+
+	pHttpDownloader->mStream = this;
+	ret = pHttpDownloader->startHttpDownload(mDownloaderMap, mState->url, parseHttpData);
+	if (ret < 0) {
+		DEBUG_TRACE(LOG_ERR, "startHttpDownload error!");
+		return -1;
+	}
+
+	return ret;
+}
+
+int AudioPDStream::pauseStream()
+{
+	AudioHttpDownloader *pHttpDownloader;
+	int ret = -1;
+
+	for (mIter = mDownloaderMap.begin(); mIter != mDownloaderMap.end(); mIter++) {
+		pHttpDownloader = (AudioHttpDownloader *)mIter->second;
+
+		ret = pHttpDownloader->pauseHttpDownload();
+		if (ret < 0) {
+			DEBUG_TRACE(LOG_ERR, "pauseHttpDownload error!");
+			return ret;
+		}
+	}
+
+	ret = mPlayer->pausePlay();
+	if (ret < 0) {
+		DEBUG_TRACE(LOG_ERR, "pausePlay error!");
+		return ret;
+	}
+
+	return ret;
+}
+
+int AudioPDStream::resumeStream()
+{
+	AudioHttpDownloader *pHttpDownloader;
+	int ret = -1;
+
+	for (mIter = mDownloaderMap.begin(); mIter != mDownloaderMap.end(); mIter++) {
+		pHttpDownloader = (AudioHttpDownloader *)mIter->second;
+
+		ret = pHttpDownloader->resumeHttpDownload();
+		if (ret < 0) {
+			DEBUG_TRACE(LOG_ERR, "resumeHttpDownload error!");
+			return ret;
+		}
+	}
+	return ret;
+}
+
+int AudioPDStream::stopStream()
+{
+	AudioHttpDownloader *pHttpDownloader;
+	int ret = -1;
+
+	for (mIter = mDownloaderMap.begin(); mIter != mDownloaderMap.end(); mIter++) {
+		pHttpDownloader = (AudioHttpDownloader *)mIter->second;
+
+		ret = pHttpDownloader->stopHttpDownload();
+		if (ret < 0) {
+			DEBUG_TRACE(LOG_ERR, "stopHttpDownload error!");
+			return ret;
+		}
+	}
+
+	ret = mPlayer->stopPlay();
+	if (ret < 0) {
+		DEBUG_TRACE(LOG_ERR, "Media stopPlay error!");
+		return ret;
+	}
+
+	return ret;
+}
+
+std::map<void *, AudioHttpDownloader *> &AudioPDStream::getDownloaderMap()
+{
+	return mDownloaderMap;
+}
+
+void AudioPDStream::removeMapAll()
+{
+	for (mIter = mDownloaderMap.begin(); mIter != mDownloaderMap.end(); mIter++) {
+		delete mIter->second;
+	}
+}
+
+int AudioPDStream::parseHttpData(void *ptr, int type, size_t len, void *p)
```

Comment:
almost same code in HASSTream already.
You can use builder(setStreamArgs(Type, Stream)  will be enough, then audiostream do that by itself.